DOC-440 K8s guides for 'Developer Guides' (#1173)
RedisLabs/redislabs-docs: content/platforms/kubernetes/tasks/ldap-on-k8s.md

---
Title: Configuring LDAP Authentication

> description: This task describes how to configure LDAP-based authentication for a cluster on Kubernetes.

Reviewer: ```suggestion
description: How to configure LDAP-based authentication for a Redis Enterprise cluster on Kubernetes.
```

weight: 90
alwaysopen: false
categories: ["Platforms"]
aliases:
---

> Throughout this tutorial we will assume there is a single namespace called ‘db'.

Reviewer: ```suggestion
This tutorial describes how to configure LDAP-based authentication for a Redis Enterprise cluster on Kubernetes. Throughout this tutorial, we will assume that there is a single namespace called ‘db'.
```

Reviewer: @alexmilowski Can you summarize the steps of the tutorial up front? There's a lot of information here, and a little roadmap would be useful.

> ## LDAP Overview

Reviewer: ```suggestion
## LDAP overview
```

At minimum, you’ll need the following information about your LDAP server:

 * The base name for where to find users (e.g., ou=users,dc=example,dc=org)
 * The filter to identify a particular user (e.g., uid=%u)

>  * The bind DN for the account that can search (e.g., cn=admin,dc=example,dc=org)

Reviewer 2: DN stands for Distinguished Name - but anyone that touches LDAP uses this nomenclature and rarely refers to it by the full name, from my experience.

 * The password for the bind DN
 * The server host and port.
 * Whether you are using SSL

In this tutorial, we will use a test server deployed on the same K8s cluster.

## Setting up an LDAP server

If you don’t already have an LDAP server, you can easily deploy one in the same K8s

> cluster for testing using a [helm chart for OpenLDAP](https://hub.helm.sh/charts/stable/openldap):

Reviewer 2: ```suggestion
cluster for testing using a [helm chart for OpenLDAP](https://github.com/helm/charts/tree/master/stable/openldap):
```

1. Create a ldap-values.yaml file that contains:

   ```
   env:

>      LDAP_ORGANISATION: "ACME Inc."

Reviewer 2: @banker This is a parameter name the helm chart expects, not a spelling choice.

     LDAP_DOMAIN: "example.org"
   adminPassword: admin
   configPassword: config
   ```

1. Deploy OpenLDAP:

   ```
   helm install ldap -f ldap-values.yaml stable/openldap
   ```

> 1. Forward the pod to your localhost so you can configure the users. The ‘xxx’ should be replace with the actual pod hash:

Reviewer 2: No. https://en.wikipedia.org/wiki/Localhost

   ```
   kubectl port-forward pod/ldap-openldap-xxx 3889:389
   ```

1. Create a file called users.ldif for the users Organizational Unit:

   ```
   dn: ou=users,dc=example,dc=org
   objectClass: organizationalUnit
   ou: users
   ```

1. Create a file called user.ldif for your test user:

   ```
   dn: uid=tester,ou=users,dc=example,dc=org
   objectClass: top
   objectClass: account
   objectClass: posixAccount
   cn: tester
   uid: tester
   uidNumber: 1001
   gidNumber: 1001
   homeDirectory: /home/tester
   userPassword: {SSHA}5IcQ2zo5wkCCohXHGWteDMBDbJElbChP
   ```

   The password for the user is “tester”. If you’d like to change it, you can with:

   ```
   ldappasswd  -h localhost -p 3889 -s newpassword -W -D "cn=admin,dc=example,dc=org" -x "uid=tester,ou=users,dc=example,dc=org"
   ```

> 1. Create the OU and user:

Reviewer 2: I don't think we should call out OU.

```
testsaslauthd -u tester -p tester
```

Once the saslauthd daemon can successfully authenticate users, you need to add the user to the list of allowed users in the Redis Enterprise cluster. You can do this through the administrative UI or via the REST API. **A user that is not added is not allowed to authenticate.** If you can authenticate with testsaslauthd but not via the REST API or the UI, then you need to verify the user has been added to the Redis Enterprise cluster.

If you want to automatically add your LDAP user, you can simply use the REST API. You’ll need to have access to the API somehow. In this example, the API has been port forwarded to the local host:

```
cat << EOF > add-user.json
{
    "name":"tester",
    "email":"tester@example.org",
    "role":"admin",
    "email_alerts":false,
    "auth_method":"external"
}
EOF
curl -v -k -u "demo@redislabs.com:xxx" -X POST -d @add-user.json -H "Content-Type: application/json" https://localhost:9443/v1/users
```

### Deploying a Redis Enterprise Cluster

> A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise CR.

Reviewer: ```suggestion
A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise custom resource.
```

> A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise CR.

Reviewer 2: Custom Resource. https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

You can create a simple test cluster that can be used with the examples below via this CR:

```
apiVersion: app.redislabs.com/v1
kind: RedisEnterpriseCluster
metadata:
  name: ldap
spec:
  nodes: 3
  redisEnterpriseNodeResources:
    limits:
      cpu: 1
      memory: 3Gi
    requests:
      cpu: 1
      memory: 3Gi
```

### LDAP Configuration

A new Redis Enterprise cluster does not have any LDAP configuration but is ready

> to accept them. The saslauthd daemon is already running and configured to use

Reviewer 2: I think you're changing the meaning here...

> LDAP as a mechanism. All you need to do is to provide the cluster

Reviewer 2: also, here

```
rladmin cluster config saslauthd_ldap_conf ldap.conf
```

The saslauthd daemon is configured to use the file `/etc/opt/redislabs/saslauthd.conf` on each node. You cannot edit this file directly on the node. Instead, you just need a local copy of the configuration you desire and rladmin command will update the cluster configuration. The cluster will update all the node’s configuration and preserve the setting after any pod restarts.

### Using the REST API to update the configuration
Alternatively, you can post the same configuration data in a request to the REST API for the cluster. First you forward the cluster API from a pod:

```
kubectl port-forward pod/ldap-0 9443
```

Then you change the LDAP configuration with a JSON version of the ldap configuration.

```
cat << EOF > tmp.json
{"saslauthd_ldap_conf": "ldap_servers: ldap://ldap-openldap.bdb.svc:389\nldap_search_base: ou=users,dc=example,dc=org\nldap_filter: (uid=%u)\nldap_bind_dn: cn=admin,dc=example,dc=org\nldap_password: admin\n"}
EOF
curl -f -k -u "demo@redislabs.com:xxx" -X PUT -d @tmp.json -H "Content-Type: application/json" https://localhost:9443/v1/cluster
```

> The key value for “saslauthd_ldap_conf” is the configuration file verbatim as it would be in the saslauthd configuration file. As such, it can be relatively tricky to get the syntax right and generating this programmatically is helpful. A little python program (make_request.py) can help:

Reviewer: Please ensure that the names of processes like saslauthd are between back ticks: `saslauthd`

```
import sys
import json

with open(sys.argv[2]) as data:
   d = {}
   d[sys.argv[1]] = data.read()
   print(json.dumps(d))
```

And so we can do:

```
python make_request.py saslauthd_ldap_conf ldap.conf > tmp.json
curl -f -k -u "demo@redislabs.com:xxx" -X PUT -d @tmp.json -H "Content-Type: application/json" https://localhost:9443/v1/cluster
```

> ### Automating LDAP Configuration

Reviewer: Please change all titles to sentence case

   ```

2. Submit this job:

   ```
   apiVersion: batch/v1
   kind: Job
   metadata:
     name: config-ldap
   spec:
     backoffLimit: 4
     template:
       spec:
         serviceAccountName: redis-enterprise-operator
         restartPolicy: Never
         containers:
           - name: config
             image: you/job-config-ldap:1
             env:
               - name: MY_NAMESPACE
                 valueFrom:
                   fieldRef:
                     fieldPath: metadata.namespace
               - name: CLUSTER_NAME
                 value: ldap
               - name: CLUSTER_USER
                 valueFrom:
                  secretKeyRef:
                    name: ldap
                    key: username
               - name: CLUSTER_PASSWORD
                 valueFrom:
                  secretKeyRef:
                    name: ldap
                    key: password
             volumeMounts:
               - name: config
                 mountPath: "/config"
                 readOnly: true
         volumes:
           - name: config
             configMap:
               name: config-job
   ```

>    In the above job, the cluster name is “ldap”. You will have to change this to the name of your cluster in all three of the CLUSTER_NAME, CLUSTER_USER, and CLUSTER_PASSWORD environment variables.

Reviewer: Can you choose a more realistic cluster name here? It may confuse the user to call the cluster "ladp"

>    In the above job, the cluster name is “ldap”. You will have to change this to the name of your cluster in all three of the CLUSTER_NAME, CLUSTER_USER, and CLUSTER_PASSWORD environment variables.

Reviewer: Please put env variables in back ticks

> This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise CR. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.

Reviewer: ```suggestion
This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise custom resource. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.
```

> This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise CR. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.

Reviewer: CR?

Reviewer: Please avoid starting sentences with "It", as the antecedent is often unclear.

Reviewer: Suggest removing "This time period is tunable". Seems irrelevant here.

> You'll need to create the docker image (i.e., in the example as "you/job-config-ldap:1")

Reviewer: ```suggestion
You'll need to create the Docker image (i.e., in the example as "you/job-config-ldap:1")
```

Reviewer: Can you be more specific about what Docker image the reader needs to create? This is the first mention of a Docker image here.

and make it available to your K8s cluster.

> You can build this docker image yourself with the following:

Reviewer: ```suggestion
You can build this Docker image yourself with the following:
```

 * `ldap_tls_cacert_dir` - a directory containing CA certificates
 * `ldap_tls_cert` - the client certificate file
 * `ldap_tls_key` - the client private key

You can provide these files and directories using a volume mount. On the cluster specification, you can specify the extra volumes via “volumes” and the extra volume mounts via “redisEnterpriseVolumeMounts”.

One way to handle the certificate files is put them into a ConfigMap and mount the volume from it. For example, if we have a special certificate authority in a file a called “ca.crt”, we can create a ConfigMap:

```
kubectl create configmap ldap-certs --from-file=ca.crt=ca.crt
```

Then in our cluster, we add the volume and volume mount:

```
apiVersion: app.redislabs.com/v1
kind: RedisEnterpriseCluster
metadata:
  name: ldap
spec:
  nodes: 3
  redisEnterpriseNodeResources:
    limits:
      cpu: 1
      memory: 3Gi
    requests:
      cpu: 1
      memory: 3Gi
  volumes:
   - name: ldap-volume
     configMap:
       name: ldap-certs
  redisEnterpriseVolumeMounts:
   - name: ldap-volume
     mountPath: /opt/ldap
```

The result is we can use the file path “/opt/ldap/ca.crt” for the “ldap_tls_cert_file” parameter.

> You can use the usual techniques to map a set of files into a directory for the “ldap_tls_cert_dir” parameter if you have multiple CA certificates:

Reviewer: Can you link "the usual techniques" to some official documentation?

    limits:
      cpu: 1
      memory: 3Gi
    requests:
      cpu: 1
      memory: 3Gi
  volumes:
   - name: ldap-volume
     configMap:
       name: ldap-certs
       items:
        - key: A.crt
          path: ca/A.crt
        - key: B.crt
          path: ca/B.crt
        - key: client.crt
          path: client.crt
        - key: client.key
          path: client.key
  redisEnterpriseVolumeMounts:
   - name: ldap-volume
     mountPath: /opt/ldap
```

which allows you to use:

```
ldap_tls_cacert_dir: /opt/ldap/ca
ldap_tls_cert: /opt/ldap/client.crt
ldap_tls_key: /opt/ldap/client.key
```

## Known Issues

1. The saslauthd daemon is not restarted when the configuration changes. If you have never configured LDAP, it will still work. The configuration file does not exist until you do and so it will get read.
Unfortunately, if you change the configuration and have previously tried to authenticate with an external user, the configuration will not get read again until the saslauthd daemon is restarted. There is currently no way to do that via K8s other than to restart the pods because K8s administrators don’t have access to root in the Redis Enterprise container.

>    You can proceed with caution and restart the pods by:

Reviewer: ```suggestion
   You can proceed with caution and restart the pods by running:
```

   ```
   kubectl rollout restart statefulset name-of-cluster
   ```

   where "name-of-cluster" is the name of the Redis Enterprise cluster in your CR.

> 1. For cluster deployed LDAP servers the suffix “cluster.local” will not resolve properly. You can remove the cluster.local suffix for LDAP deployments in the same K8s cluster.

Reviewer: ```suggestion
1. For cluster-deployed LDAP servers, the suffix “cluster.local” will not resolve properly. You can remove the "cluster.local" suffix for LDAP deployments in the same K8s cluster.
```

Reviewer: You can remove or you should remove?
